composer/trainer/devices/device_tpu.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+"""The TPU device used for training."""
+
+from __future__ import annotations
+
+import logging
+from contextlib import contextmanager
+from typing import Generator, TypeVar, Union
+
+import torch
+
+from composer.core.precision import Precision
+from composer.trainer.devices.device import Device, T_nnModule
+import torch_xla
+
+logger = logging.getLogger(__name__)
+
+__all__ = ["DeviceTPU"]
+
+T_nnModule = TypeVar("T_nnModule", bound=torch.nn.Module)
+
+
+class DeviceTPU(Device):
+    """An extension of :class:`~composer.trainer.devices.device.Device` for TPUs
+
+    This class takes no arguments.
+    """
+
+    def __init__(self):
+        import torch_xla.core.xla_model as xm
+
+        self._device = xm.xla_device()
+        
+    def module_to_device(self, module: T_nnModule) -> T_nnModule:
+        import torch_xla.distributed.xla_multiprocessing as xmp
+        
+        WRAPPED_MODEL = xmp.MpModelWrapper(module)
+        return WRAPPED_MODEL.to(self._device)
+
+    def tensor_to_device(self, tensor: torch.Tensor) -> torch.Tensor:
+        return tensor.to(self._device)
+
+    @contextmanager
+    def precision_context(self, precision: Union[str, Precision]) -> Generator[None, None, None]:
```

Comment:
The `precision_context` was removed; see https://github.com/mosaicml/composer/blob/dev/composer/core/precision.py#L35